### PR TITLE
feat(graphqlmetrics): enable prometheus metrics

### PR DESCRIPTION
**Pull request by AndreasZeissner**
## Motivation and Context

This PR introduces exposing metrics on the graphqlmetrics service.
The prometheus export can be enabled by using the following environment variables:

`PROMETHEUS_ENABLED`: enable or disable exposing metrics defaults to false
`PROMETHEUS_LISTEN_ADDR`: the listen address for the prometheus server defaults to 127.0.0.1:8088 to exposing metrics to the public bt accident
`PROMETHEUS_PATH`: the path to expose the metrics on defaults to /metrics

## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: graphqlmetrics/core/server.go [hunk 1]
```diff
@@ -39,16 +44,47 @@ func NewServer(metricsService graphqlmetricsv1connect.GraphQLMetricsServiceHandl
...
-	}))
+	})
+
+	if s.metricConfig.OpenTelemetry.Enabled {
+		tp, err := s.metricConfig.NewTracerProvider()
+		if err != nil {
+			s.logger.Error("Error creating tracing provider", zap.Error(err))
+		}
+		defer func() {
+			if err := tp.Shutdown(context.Background()); err != nil {
```
Comment (StarpTech) [inline graphqlmetrics/core/server.go:63]: Shutdown logic should be done in `Shutdown` with a proper context.

File: graphqlmetrics/core/server.go [hunk 2]
```diff
@@ -39,16 +44,47 @@ func NewServer(metricsService graphqlmetricsv1connect.GraphQLMetricsServiceHandl
...
 	mux.Handle(path, authenticate(s.jwtSecret, s.logger, handler))
 
+	if s.metricConfig.Prometheus.Enabled {
+		mp, registry, err := s.metricConfig.NewPrometheusMeterProvider()
+		if err != nil {
+			s.logger.Error("Failed to create Prometheus exporter", zap.Error(err))
+		}
+
+		defer func() {
+			if err := mp.Shutdown(context.Background()); err != nil {
```
Comment (StarpTech) [inline graphqlmetrics/core/server.go:80]: Same here.

File: graphqlmetrics/core/server.go [hunk 3]
```diff
@@ -84,6 +120,30 @@ func (s *Server) Shutdown(ctx context.Context) error {
 	return nil
 }
 
+func (s *Server) ShutdownPrometheusServer(ctx context.Context) error {
```
Comment (StarpTech) [inline graphqlmetrics/core/server.go:123]: Call them in `Shutdown` and make them private.

File: graphqlmetrics/core/telemetry_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		// as the server setup is all encapsulated within main,
+		// at least check, if the new code is covered and if the
+		// handlers for the prometheus server are registered
+		go func() {
+			if err := tt.svr.StartPrometheusServer(); err != nil {
```
Comment (StarpTech) [inline graphqlmetrics/core/telemetry_test.go:64]: Please add an integration test that actually test the server implementation. Additionally, we want to validate in the tests that proper metric datapoints has been exposed.

```go
    // 1. Setup the server
	svr := core.NewServer(
		ms,
	   core.WithMetrics(metricsConfig),
		core.WithListenAddr(cfg.ListenAddr),
	)

// 2. Make a request to the buf service
// 3. Validate metrics on the prometheus endpoint
```

File: graphqlmetrics/pkg/telemetry/metrics_prometheus.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	// Only available on Linux and Windows systems
+	registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+
+	promExporter, err := otelprom.New(
+		otelprom.WithoutUnits(),
+		otelprom.WithRegisterer(registry),
+	)
+
+	if err != nil {
+		panic(err)
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/metrics_prometheus.go:66]: Don't use panic for err handling. Propagate the err so it is possible to handle it and to add more context.

File: graphqlmetrics/pkg/telemetry/tracing.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+)
+
+const (
+	// WithEndpoint respects the following environment variables:
+	// OTEL_EXPORTER_OTLP_ENDPOINT
+	// OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
+	DefaultGrpcTelemetryEndpoint = "cosmo-otel.wundergraph.com"
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/tracing.go:20]: This endpoint is intended for customers routers. We should not give the impression that we can use it for platform monitoring.

File: graphqlmetrics/pkg/telemetry/tracing.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func (c *Config) NewTracerProvider() (*sdktrace.TracerProvider, error) {
+	exporter, err := newTraceExporter()
+	if err != nil {
+		return nil, err
+	}
+
+	resources, err := resource.New(
+		context.Background(),
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/tracing.go:51]: `context.Background()` is rarely used. In most of the cases, you should allow to pass a custom context to allow cancellations.

File: graphqlmetrics/pkg/telemetry/metrics_prometheus.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(promExporter),
+		sdkmetric.WithResource(resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(c.Name),
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/metrics_prometheus.go:73]: Please also track the version with `	resource.WithAttributes(semconv.ServiceVersionKey.String(c.Version))`

File: graphqlmetrics/pkg/telemetry/tracing.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+)
+
+const (
+	// WithEndpoint respects the following environment variables:
+	// OTEL_EXPORTER_OTLP_ENDPOINT
+	// OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
+	DefaultGrpcTelemetryEndpoint = "cosmo-otel.wundergraph.com"
+)
+
+func buildResourceOptions() resource.Option {
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/tracing.go:23]: We don't have to duplicate those information when they are already cover above with standardized attributes.

File: graphqlmetrics/pkg/telemetry/tracing.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+func newOtlpTraceGrpcClient() otlptrace.Client {
+	return otlptracegrpc.NewClient(
+		otlptracegrpc.WithEndpoint(DefaultGrpcTelemetryEndpoint),
+		otlptracegrpc.WithCompressor("gzip"),
+	)
+}
+
+func newTraceExporter() (*otlptrace.Exporter, error) {
+	return otlptrace.New(
+		context.Background(),
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/tracing.go:39]: Same here with the context.

File: graphqlmetrics/pkg/telemetry/metrics_prometheus.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		otelprom.WithRegisterer(registry),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(promExporter),
+		sdkmetric.WithResource(resource.NewWithAttributes(
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/metrics_prometheus.go:71]: Please attach those resource attribute presets as well

```go
	resource.WithProcessPID(),
		resource.WithOSType(),
		resource.WithTelemetrySDK(),
		resource.WithHost(),
```

File: graphqlmetrics/pkg/telemetry/metrics_prometheus.go [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(c.Name),
+		)),
+	)
+
+	otel.SetMeterProvider(mp)
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/metrics_prometheus.go:77]: This will set a global meter. This can be an issue with concurrent tests. We only work with dedicated meters.

File: graphqlmetrics/core/server.go
Comment (StarpTech) [inline graphqlmetrics/core/server.go:1]: Please add at least a metric that adds more information than the process stats. A good metric could be the total amount of requests for every customer.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: graphqlmetrics/pkg/telemetry/tracing.go (hunk 9 above)
Comment (AndreasZeissner) [reply graphqlmetrics/pkg/telemetry/tracing.go:23]: I would rather remove the tracing related implementations for now and bring them in, when tracing is integrated.

File: graphqlmetrics/core/server.go (hunk 1 above)
Comment (AndreasZeissner) [reply graphqlmetrics/core/server.go:63]: Would it be ok to use the same context as here as the server itself?

File: graphqlmetrics/core/telemetry_test.go (hunk 4 above)
Comment (AndreasZeissner) [reply graphqlmetrics/core/telemetry_test.go:64]: As far as I see, in graphqlmetrics, there are only Integration Tests.

Is this ok to only have those here e.g. mocking the MetricsService seemed odd to me in the first place already to get some coverage here.

File: graphqlmetrics/core/server.go (hunk 1 above)
Comment (AndreasZeissner) [reply graphqlmetrics/core/server.go]: I moved the shutdown together, seems to work fine to reuse the context.

Let me know if this might be an issue.

File: graphqlmetrics/pkg/telemetry/tracing.go (hunk 7 above)
Comment (AndreasZeissner) [reply graphqlmetrics/pkg/telemetry/tracing.go]: I removed tracing related references for now to split up metrics and tracing properly.

File: graphqlmetrics/pkg/telemetry/interceptor.go [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+	WgOperationName     = attribute.Key("wg.operation.name")
+	WgOperationMethod   = attribute.Key("wg.operation.method")
+	WgOperationProtocol = attribute.Key("wg.operation.protocol")
+)
+
+func defaultAttributes(req connect.AnyRequest) []attribute.KeyValue {
+	var attributes []attribute.KeyValue
+
+	attributes = append(attributes, WgOperationName.String(req.Spec().Procedure))
+	attributes = append(attributes, WgOperationMethod.String(req.HTTPMethod()))
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/interceptor.go:20]: There are standardized attributes for basic http/net attributes.

File: graphqlmetrics/pkg/telemetry/config.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+// Config represents the configuration for the agent.
+type Config struct {
+	// Name represents the service name for metrics. The default value is cosmo-router.
+	Name string
+	// Version represents the service version for metrics. The default value is dev.
+	Version string
+	// Prometheus includes the Prometheus configuration
+	Prometheus PrometheusConfig
+	// CustomPrometheusMetrics to be collected
+	CustomMetrics CustomMetrics
```
Comment (StarpTech) [inline graphqlmetrics/pkg/telemetry/config.go:42]: Is this for testing?

File: graphqlmetrics/core/server.go [hunk 15]
```diff
@@ -32,21 +44,53 @@ func NewServer(metricsService graphqlmetricsv1connect.GraphQLMetricsServiceHandl
...
 	)
-	mux.Handle("/health", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+	healthHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(http.StatusOK)
-	}))
+	})
+
+	mux.Handle("/health", healthHandler)
 	mux.Handle(path, authenticate(s.jwtSecret, s.logger, handler))
```
Comment (StarpTech) [inline graphqlmetrics/core/server.go:94]: Unfortunately, I have seen this too late. Instead of using an interceptor we can track the metrics in a regular handler. We would also get access to the JWT which contains the organization and federated graph id. That would make the metrics much more insightful.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: graphqlmetrics/core/telemetry_test.go [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+				err := svr.Shutdown(ctx)
+				if err != nil {
+					t.Fatalf("Failed to shut down server: %v", err)
+				}
+			}()
+			time.Sleep(2 * time.Second)
+
+			endpoint := fmt.Sprintf("http://%s%s", tt.prom.ListenAddr, tt.prom.Path)
+			resp, err := http.Get(endpoint)
+
```
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply graphqlmetrics/core/telemetry_test.go:111]: Calling it earlier here is not possible as resp could be nil.
I could change the test and split it up in two tests. 

The intention here was to express the two possibilities, either there will be a response, or it will be nil as there was no connection possible.

File: graphqlmetrics/core/server.go [hunk 17]
```diff
@@ -60,13 +104,24 @@ func (s *Server) bootstrap() {
 	}
 }
 
-func (s *Server) Start() error {
+func (s *Server) Start(stop context.CancelFunc) error {
```
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply graphqlmetrics/core/server.go:107]: What would be the way to go here?
The cancel function is only called within the main servers go routine. Even when the prometheus listener dies the context of the main server is not cancelled, isn't it?

File: graphqlmetrics/pkg/telemetry/config.go (hunk 14 above)
Comment (AndreasZeissner) [reply graphqlmetrics/pkg/telemetry/config.go:41]: I am not sure what you mean here.

I updated some comments, it's not reflected here in the PR view, I hope this makes more sense:

https://github.com/wundergraph/cosmo/commit/bd2a30ff4c577f373677a49888a398bddd66a961

File: graphqlmetrics/core/server.go (hunk 15 above)
Comment (AndreasZeissner) [reply graphqlmetrics/core/server.go:94]: Wouldn't it be possible to pass the parent context to the interceptor and the pull out the same information?

File: graphqlmetrics/core/telemetry_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply graphqlmetrics/core/telemetry_test.go:112]: You're right. Let me rephrase.

```go
if err != nil {
// err handling
}
defer resp.Body.Close()

// response handling
```

File: graphqlmetrics/core/server.go (hunk 15 above)
Comment (StarpTech) [reply graphqlmetrics/core/server.go:94]: Yes, when you get access to the request context.

File: graphqlmetrics/core/telemetry_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply graphqlmetrics/core/telemetry_test.go:112]: The test case does not allow this as in once case resp can be nil and the error branch is a valid case here. 

"prometheus was not enabled and no response should have been able to be made and is therefore nil". 

I get your point, I moved the defer to the earliest point, where I know, there is something to close. I will let that sink in and think of splitting it up in two tests.

[the remark this answers is not in the document]
Comment (StarpTech) [reply graphqlmetrics/core/telemetry_test.go:112]: Ok

File: graphqlmetrics/core/server.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply graphqlmetrics/core/server.go:107]: Just don't pass the cancelFunc here and handle at the place where Start is called.

File: graphqlmetrics/pkg/telemetry/interceptor.go (hunk 13 above)
Comment (AndreasZeissner) [reply graphqlmetrics/pkg/telemetry/interceptor.go]: I updated them to use the GRPC Connect specific one, you can see this in the tests.
I would say this makes sense now.

File: graphqlmetrics/core/server.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply graphqlmetrics/core/server.go]: Thanks.

File: graphqlmetrics/core/telemetry_test.go [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+					Prometheus: tt.prom,
+				}))
+
+			go func() {
+				if err := svr.Start(stop); err != nil && !errors.Is(err, http.ErrServerClosed) {
+					t.Logf("failed starting main server")
+				}
+			}()
+
+			defer func() {
```
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply graphqlmetrics/core/telemetry_test.go:103]: Unfortunately this seems not to cleanup the server reliable. 

I would leave it like it is as it keeps the test run predictable even when restarting the server in between runs.

File: graphqlmetrics/core/server.go (hunk 15 above)
Comment (AndreasZeissner) [reply graphqlmetrics/core/server.go:94]: Works with the interceptor. 

The claims could be unset, this shouldn't happen as the  validation happens already before, but in case the claims are empty, I pass back a `INVALID_ARGUMENT`. Let me know about your opinion.

Comment (StarpTech) [reply graphqlmetrics/core/server.go:94]: Sounds good

File: graphqlmetrics/core/telemetry_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply graphqlmetrics/core/telemetry_test.go:103]: ok

**Review by StarpTech — APPROVED** [review]
LGTM
